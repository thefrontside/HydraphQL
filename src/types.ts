import { getDirective } from "@graphql-tools/utils";
import type DataLoader from "dataloader";
import type {
  GraphQLFieldConfig,
  GraphQLNamedType,
  GraphQLObjectType,
  GraphQLSchema,
} from "graphql";
import type { Application, Module } from "graphql-modules";

export interface NodeQuery {
  ref?: string;
  args?: Record<string, unknown>;
}

export interface NodeId {
  source: string;
  typename: string;
  query: NodeQuery;
}

export type BatchLoadFn<Context extends GraphQLContext> = (
  keys: readonly NodeQuery[],
  context: Context,
) => PromiseLike<ArrayLike<unknown>>;

export interface GraphQLContext {
  application: Application;
}

export interface ResolverContext extends GraphQLContext {
  loader: DataLoader<string, unknown>;
}

export type OmitFirst<T extends unknown[]> = T extends [
  x: unknown,
  ...args: infer R,
]
  ? R
  : [];

export interface DirectiveMapperAPI {
  getImplementingTypes: (interfaceName: string) => GraphQLObjectType[];
  getDirective: (
    ...args: OmitFirst<Parameters<typeof getDirective>>
  ) => ReturnType<typeof getDirective>;
  typeMap: Partial<Record<string, GraphQLNamedType>>;
}

export type FieldDirectiveMapper = (
  fieldName: string,
  field: GraphQLFieldConfig<
    { id: string },
    ResolverContext,
    Record<string, unknown> | undefined
  >,
  directive: Record<string, unknown>,
  api: DirectiveMapperAPI & { typeName: string },
) => void;

export interface NamedType {
  implements: string | null;
  discriminates: Set<string>;
}

export interface GraphQLModule {
  mappers?: Record<string, FieldDirectiveMapper>;
  postTransform?: (schema: GraphQLSchema) => GraphQLSchema;
  module: Module;
}
