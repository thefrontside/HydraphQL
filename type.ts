void (async () => {
  const { mkdir, readFile, writeFile } = await import("fs/promises");

  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
  const packageJson = JSON.parse(await readFile("./package.json", "utf-8"));
  const type = process.argv[2];

  if (!["module", "commonjs"].includes(type)) {
    throw new Error(`Invalid type: ${type}`);
  }

  await writeFile(
    "./package.json",
    JSON.stringify({ ...packageJson, type }, null, 2),
  );

  const dir = type === "module" ? "esm" : "cjs";

  await mkdir(`./dist/${dir}`, { recursive: true });
  await writeFile(
    `./dist/${dir}/package.json`,
    JSON.stringify({ type }, null, 2),
  );
})();
