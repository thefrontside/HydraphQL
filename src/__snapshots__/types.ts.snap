/* eslint-disable */
// @ts-nocheck
import { GraphQLResolveInfo, GraphQLScalarType, GraphQLScalarTypeConfig } from 'graphql';
export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
export type MakeEmpty<T extends { [key: string]: unknown }, K extends keyof T> = { [_ in K]?: never };
export type Incremental<T> = T | { [P in keyof T]?: P extends ' $fragmentName' | '__typename' ? T[P] : never };
export type RequireFields<T, K extends keyof T> = Omit<T, K> & { [P in K]-?: NonNullable<T[P]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: { input: string; output: string; }
  String: { input: string; output: string; }
  Boolean: { input: boolean; output: boolean; }
  Int: { input: number; output: number; }
  Float: { input: number; output: number; }
  _DirectiveArgument_: { input: any; output: any; }
};

export type Node = {
  id: Scalars['ID']['output'];
};

export type Query = {
  __typename?: 'Query';
  node?: Maybe<Node>;
  nodes: Array<Maybe<Node>>;
};


export type QueryNodeArgs = {
  id: Scalars['ID']['input'];
};


export type QueryNodesArgs = {
  ids: Array<Scalars['ID']['input']>;
};



export type ResolverTypeWrapper<T> = Promise<T> | T;


export type ResolverWithResolve<TResult, TParent, TContext, TArgs> = {
  resolve: ResolverFn<TResult, TParent, TContext, TArgs>;
};
export type Resolver<TResult, TParent = {}, TContext = {}, TArgs = {}> = ResolverFn<TResult, TParent, TContext, TArgs> | ResolverWithResolve<TResult, TParent, TContext, TArgs>;

export type ResolverFn<TResult, TParent, TContext, TArgs> = (
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => Promise<TResult> | TResult;

export type SubscriptionSubscribeFn<TResult, TParent, TContext, TArgs> = (
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => AsyncIterable<TResult> | Promise<AsyncIterable<TResult>>;

export type SubscriptionResolveFn<TResult, TParent, TContext, TArgs> = (
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => TResult | Promise<TResult>;

export interface SubscriptionSubscriberObject<TResult, TKey extends string, TParent, TContext, TArgs> {
  subscribe: SubscriptionSubscribeFn<{ [key in TKey]: TResult }, TParent, TContext, TArgs>;
  resolve?: SubscriptionResolveFn<TResult, { [key in TKey]: TResult }, TContext, TArgs>;
}

export interface SubscriptionResolverObject<TResult, TParent, TContext, TArgs> {
  subscribe: SubscriptionSubscribeFn<any, TParent, TContext, TArgs>;
  resolve: SubscriptionResolveFn<TResult, any, TContext, TArgs>;
}

export type SubscriptionObject<TResult, TKey extends string, TParent, TContext, TArgs> =
  | SubscriptionSubscriberObject<TResult, TKey, TParent, TContext, TArgs>
  | SubscriptionResolverObject<TResult, TParent, TContext, TArgs>;

export type SubscriptionResolver<TResult, TKey extends string, TParent = {}, TContext = {}, TArgs = {}> =
  | ((...args: any[]) => SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>)
  | SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>;

export type TypeResolveFn<TTypes, TParent = {}, TContext = {}> = (
  parent: TParent,
  context: TContext,
  info: GraphQLResolveInfo
) => Maybe<TTypes> | Promise<Maybe<TTypes>>;

export type IsTypeOfResolverFn<T = {}, TContext = {}> = (obj: T, context: TContext, info: GraphQLResolveInfo) => boolean | Promise<boolean>;

export type NextResolverFn<T> = () => Promise<T>;

export type DirectiveResolverFn<TResult = {}, TParent = {}, TContext = {}, TArgs = {}> = (
  next: NextResolverFn<TResult>,
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => TResult | Promise<TResult>;


/** Mapping of interface types */
export type ResolversInterfaceTypes<RefType extends Record<string, unknown>> = {
  Node: never;
};

/** Mapping between all available schema types and the resolvers types */
export type ResolversTypes = {
  Boolean: ResolverTypeWrapper<Scalars['Boolean']['output']>;
  ID: ResolverTypeWrapper<Scalars['ID']['output']>;
  Node: ResolverTypeWrapper<ResolversInterfaceTypes<ResolversTypes>['Node']>;
  Query: ResolverTypeWrapper<{}>;
  String: ResolverTypeWrapper<Scalars['String']['output']>;
  _DirectiveArgument_: ResolverTypeWrapper<Scalars['_DirectiveArgument_']['output']>;
};

/** Mapping between all available schema types and the resolvers parents */
export type ResolversParentTypes = {
  Boolean: Scalars['Boolean']['output'];
  ID: Scalars['ID']['output'];
  Node: ResolversInterfaceTypes<ResolversParentTypes>['Node'];
  Query: {};
  String: Scalars['String']['output'];
  _DirectiveArgument_: Scalars['_DirectiveArgument_']['output'];
};

export type DiscriminatesDirectiveArgs = {
  opaqueType?: Maybe<Scalars['String']['input']>;
  with?: Maybe<Scalars['_DirectiveArgument_']['input']>;
};

export type DiscriminatesDirectiveResolver<Result, Parent, ContextType = any, Args = DiscriminatesDirectiveArgs> = DirectiveResolverFn<Result, Parent, ContextType, Args>;

export type DiscriminationAliasDirectiveArgs = {
  type: Scalars['String']['input'];
  value: Scalars['String']['input'];
};

export type DiscriminationAliasDirectiveResolver<Result, Parent, ContextType = any, Args = DiscriminationAliasDirectiveArgs> = DirectiveResolverFn<Result, Parent, ContextType, Args>;

export type FieldDirectiveArgs = {
  at: Scalars['_DirectiveArgument_']['input'];
  default?: Maybe<Scalars['_DirectiveArgument_']['input']>;
};

export type FieldDirectiveResolver<Result, Parent, ContextType = any, Args = FieldDirectiveArgs> = DirectiveResolverFn<Result, Parent, ContextType, Args>;

export type ImplementsDirectiveArgs = {
  interface: Scalars['String']['input'];
};

export type ImplementsDirectiveResolver<Result, Parent, ContextType = any, Args = ImplementsDirectiveArgs> = DirectiveResolverFn<Result, Parent, ContextType, Args>;

export type ResolveDirectiveArgs = {
  at?: Maybe<Scalars['_DirectiveArgument_']['input']>;
  from?: Maybe<Scalars['String']['input']>;
};

export type ResolveDirectiveResolver<Result, Parent, ContextType = any, Args = ResolveDirectiveArgs> = DirectiveResolverFn<Result, Parent, ContextType, Args>;

export type NodeResolvers<ContextType = any, ParentType extends ResolversParentTypes['Node'] = ResolversParentTypes['Node']> = {
  __resolveType: TypeResolveFn<null, ParentType, ContextType>;
  id?: Resolver<ResolversTypes['ID'], ParentType, ContextType>;
};

export type QueryResolvers<ContextType = any, ParentType extends ResolversParentTypes['Query'] = ResolversParentTypes['Query']> = {
  node?: Resolver<Maybe<ResolversTypes['Node']>, ParentType, ContextType, RequireFields<QueryNodeArgs, 'id'>>;
  nodes?: Resolver<Array<Maybe<ResolversTypes['Node']>>, ParentType, ContextType, RequireFields<QueryNodesArgs, 'ids'>>;
};

export interface _DirectiveArgument_ScalarConfig extends GraphQLScalarTypeConfig<ResolversTypes['_DirectiveArgument_'], any> {
  name: '_DirectiveArgument_';
}

export type Resolvers<ContextType = any> = {
  Node?: NodeResolvers<ContextType>;
  Query?: QueryResolvers<ContextType>;
  _DirectiveArgument_?: GraphQLScalarType;
};

export type DirectiveResolvers<ContextType = any> = {
  discriminates?: DiscriminatesDirectiveResolver<any, any, ContextType>;
  discriminationAlias?: DiscriminationAliasDirectiveResolver<any, any, ContextType>;
  field?: FieldDirectiveResolver<any, any, ContextType>;
  implements?: ImplementsDirectiveResolver<any, any, ContextType>;
  resolve?: ResolveDirectiveResolver<any, any, ContextType>;
};

export type _DirectiveArgument_ = Scalars["_DirectiveArgument_"];